app/src/main/java/com/github/warnastrophy/core/data/repository/ContactsRepositoryLocal.kt

```diff
@@ -0,0 +1,159 @@
+package com.github.warnastrophy.core.data.repository
+
+import android.content.Context
+import android.util.Log
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.core.edit
+import androidx.datastore.preferences.core.stringPreferencesKey
+import androidx.datastore.preferences.preferencesDataStore
+import com.github.warnastrophy.core.data.local.StorageException
+import com.github.warnastrophy.core.model.Contact
+import com.github.warnastrophy.core.util.CryptoUtils
+import com.google.gson.Gson
+import java.util.UUID
+import kotlin.collections.iterator
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.map
+
+const val CONTACTS_DATASTORE_NAME = "contacts_encrypted"
+
+val gson = Gson()
+
+val Context.contactDataStore: DataStore<Preferences> by
+    preferencesDataStore(name = CONTACTS_DATASTORE_NAME)
+
+class ContactsRepositoryLocal(private val dataStore: DataStore<Preferences>) : ContactsRepository {
+  fun keyFor(contact: Contact) = stringPreferencesKey(contact.id)
+
+  override fun getNewUid(): String {
+    return UUID.randomUUID().toString()
+  }
+
+  override suspend fun addContact(contact: Contact): Result<Unit> {
+    var r = Result.success(Unit)
+
+    try {
+      dataStore.edit {
+        if (it.contains(keyFor(contact))) {
+          r =
+              Result.failure(
+                  StorageException.DataStoreError(
+                      Exception("Contact ${contact.id} already exists")))
+          return@edit
+        }
+
+        val ciphered = CryptoUtils.encrypt(gson.toJson(contact))
+        it[keyFor(contact)] = ciphered
+        r = Result.success(Unit)
+      }
+    } catch (e: Exception) {
+      Log.e("ContactStorage", "Error saving contact ${contact.id}", e)
+      r = Result.failure(StorageException.DataStoreError(e))
+    }
+
+    return r
+  }
+
+  override suspend fun getAllContacts(): Result<List<Contact>> {
+    val prefs = dataStore.data.first()
+    val contacts = mutableListOf<Contact>()
+    for ((_, value) in prefs.asMap()) {
+      try {
+        val decrypted = CryptoUtils.decrypt(value as String)
+        val contact = gson.fromJson(decrypted, Contact::class.java)
+        contacts.add(contact)
+      } catch (e: Exception) {
+        throw StorageException.DataStoreError(e)
+      }
+    }
+    return Result.success(contacts)
```

Comment:
I suggest this because it always return a `Result`. In current version, `StorageException.DataStoreError(e)` can be thrown and `Result` won't be return
```suggestion
    val prefs = dataStore.data.first()
    prefs.asMap().mapNotNull { (_, value) ->
        val stringValue = value as? String ?: return@mapNotNull null
        val decrypted = CryptoUtils.decrypt(stringValue)
        gson.fromJson(decrypted, Contact::class.java)
    }
}.mapFailure {
    StorageException.DataStoreError(it)
```